lifecycle-operator/controllers/common/schedulinggateshandler.go

```diff
@@ -0,0 +1,110 @@
+package common
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	klcv1alpha3 "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3"
+	apicommon "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3/common"
+	controllererrors "github.com/keptn/lifecycle-toolkit/lifecycle-operator/controllers/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate moq -pkg fake -skip-ensure -out ./fake/schedulinggateshandler_mock.go . ISchedulingGatesHandler
+type ISchedulingGatesHandler interface {
+	RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error
+	IsSchedulingEnabled() bool
+}
+
+type RemoveGatesFunc func(ctx context.Context, c client.Client, log logr.Logger, podName string, podNamespace string) error
+type GetPodsFunc func(ctx context.Context, c client.Client, log logr.Logger, ownerUID types.UID, ownerKind string, namespace string) ([]string, error)
+
+type SchedulingGatesHandler struct {
+	client.Client
+	logr.Logger
+	enabled     bool
+	removeGates RemoveGatesFunc
+	getPods     GetPodsFunc
+}
+
+func NewSchedulingGatesHandler(c client.Client, l logr.Logger, enabled bool) *SchedulingGatesHandler {
+	return &SchedulingGatesHandler{
+		Client:      c,
+		Logger:      l,
+		enabled:     enabled,
+		removeGates: removePodGates,
+		getPods:     getPodsOfOwner,
+	}
+}
+
+func (h *SchedulingGatesHandler) RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error {
+	switch workloadInstance.Spec.ResourceReference.Kind {
+	case "Pod":
+		return h.removeGates(ctx, h.Client, h.Logger, workloadInstance.Spec.ResourceReference.Name, workloadInstance.Namespace)
+	case "ReplicaSet", "StatefulSet", "DaemonSet":
+		podList, err := h.getPods(ctx, h.Client, h.Logger, workloadInstance.Spec.ResourceReference.UID, workloadInstance.Spec.ResourceReference.Kind, workloadInstance.Namespace)
+		if err != nil {
+			h.Logger.Error(err, "cannot get pods")
+			return err
+		}
+		for _, pod := range podList {
+			err := h.removeGates(ctx, h.Client, h.Logger, pod, workloadInstance.Namespace)
+			if err != nil {
+				h.Logger.Error(err, "cannot remove gates from pod")
+				return err
+			}
+		}
+	default:
+		return controllererrors.ErrUnsupportedWorkloadInstanceResourceReference
+	}
+
+	return nil
+}
+
+func (h *SchedulingGatesHandler) IsSchedulingEnabled() bool {
+	return h.enabled
+}
+
+func removePodGates(ctx context.Context, c client.Client, log logr.Logger, podName string, podNamespace string) error {
```

Comment:
i think there is no need to also pass the logger, as we log any error in the `RemoveGates` function above anyway

Reply:
adapted